igvc_navigation/include/igvc_navigation/PriorityQueue.h

```diff
@@ -0,0 +1,216 @@
+                                ||       ||
+                                ||       ||
+                            ||_______________||
+
+                     yes, some of the ascii characters
+                               are ornaments
+
+                                                - Alejandro
+*/
+
+#ifndef PRIORITYQUEUE_H
+#define PRIORITYQUEUE_H
+
+#include <algorithm>
+#include <functional>
+#include <set>
+#include <sstream>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "Node.h"
+
+/**
+  key defined as <f1(s), f2(s)>
+  where...
+  f1(s) = min(g(s), rhs(s)) + h(s_start, s) + K_M
+  f2(s)min(g(s), rhs(s))
+*/
+struct Key
+{
+  float f1;
+  float f2;
+
+  Key(float f1, float f2)
+  {
+    this->f1 = f1;
+    this->f2 = f2;
+  }
+
+  Key(std::tuple<float, float> k)
+  {
+    this->f1 = std::get<0>(k);
+    this->f2 = std::get<1>(k);
+  }
+
+  // overloaded assignment operator
+  Key& operator=(const Key& other)
+  {
+    this->f1 = other.f1;
+    this->f2 = other.f2;
+
+    return *this;
+  }
+
+  // lexicograhpic ordering of keys
+  bool operator<(const Key& other) const
+  {
+    if (this->f1 < other.f1)
+      return true;
+    else if (this->f1 == other.f1)
+      return this->f2 < other.f2;
+    else
+      return false;
+  }
+
+  bool operator<=(const Key& other) const
+  {
+    if (this->f1 < other.f1)
+      return true;
+    else if (this->f1 == other.f1)
+      return this->f2 <= other.f2;
+    else
+      return false;
+  }
+
+  // Keys equal if their corresponding values are equal
+  bool operator==(const Key& other) const
+  {
+    return (this->f1 == other.f1) && (this->f2 == other.f2);
+  }
+
+  bool operator!=(const Key& other) const
+  {
+    return !(*this==other); 
+  }
+};
+
+class PriorityQueue
+{
+public:
+  /**
+  Default no-arg constructor for priority queue. Initializes priority queue
+  with compFunctor Comparator. This comparator sorts the priority queue based
+  on the lexicographic ordering of the key values.
+  */
+  PriorityQueue();
+  /**
+  Checks whether or not the priority queue contains a state with the specified
+  index.
+
+  @param[in] ind tuple representing the indices of the state
+  @return whether or not the priority queue contains an entry for that state
+  */
+  bool contains(Node n);  // O(n)
+  /**
+  Insert a state into the priority queue.
+
+  @param[in] item state to insert into the priority queue
+  */
+  void insert(Node n, Key k);  // O(n)
+  /**
+  Clears the priority queue's contents.
+  */
+  void clear();
+  /**
+  Remove an item from the priority queue for the specified node.
+
+  @param[in] n node to remove from the priority queue
+  @return whether such an item was found and removed
+  */
+  bool remove(Node n);  // O(n)
+  /**
+  Pops the top (least-cost) state from the priority queue. This has the effect of
+  reducing the size of the priority queue by one. If the priority queue is
+  empty, nothing happens.
+  */
+  void pop();  // O(1)
```

Comment:
This is O(log(n)), you're using a set, which uses a red-black tree as the backing data structure.

Reply:
:+1: I'll update

Reply:
```C++
auto it = this->find(n);

if (it == pq.end())  // no such item exists
  return false;

pq.erase(it);
return true;
```

and 

```C++
std::set<std::pair<Node, Key>>::iterator PriorityQueue::find(Node n)
{
  std::set<std::pair<Node, Key>>::iterator it = std::find_if(
      pq.begin(), pq.end(), [n](std::pair<Node, Key> const& e) { return e.first.getIndex() == n.getIndex(); });

  return it;
}
```